notebook/secciones/estructuras/link_cut_tree.cpp

```diff
@@ -0,0 +1,118 @@
+const int N_DEL = 0, N_VAL = 0; // neutral elem for delta, value
+inline int update(int x, int y){ return x+y; } // update 
+inline int get(int lval, int rval){ return lval + rval; } // query
+inline int segm_delta(int d, int len){ return d==N_DEL ? N_DEL : d*len; } // how to calc the delta on a segment
+inline int update_delta(int d1, int d2){ // update delta
+	if(d1==N_DEL) return d2; 
+	if(d2==N_DEL) return d1;
+	return update(d1, d2);
+}
+inline int update_node(int v, int d){ // update node value
+	return d==N_DEL ? v : update(v, d);
+}
+
+// splay tree
+struct node_t {
+	int sz, n_val, t_val, d;
+	bool rev;
+	node_t *c[2], *p;
+	node_t(int v_) : n_val(v_), t_val(v_), d(N_DEL), rev(0), p(0){
+		sz = 1;
+		c[0]=c[1]=0;
+	}
+	bool is_root(){ return !p || (p->c[0] != this && p->c[1] != this); }
+	void push(){
+		if(rev){
+			rev=0; swap(c[0], c[1]);
+			forr(x,0,2) if(c[x]) c[x]->rev ^= 1;
+		}
+		n_val = update_node(n_val, d); t_val = update_node(t_val, segm_delta(d, sz));
+		forr(x,0,2) if(c[x]) c[x]->d = update_delta(c[x]->d, d);
+		d = N_DEL;
+	}
+	void upd();
+};
+typedef node_t* node;
+int get_sz(node r){ return r ? r->sz : 0; }
+int get_tree_val(node r){
+	return r ? update_node(r->t_val, segm_delta(r->d, r->sz)) : N_VAL;
+}
+void node_t::upd(){
+  t_val = get(get(get_tree_val(c[0]), update_node(n_val, d)), get_tree_val(c[1]));
+  sz = 1 + get_sz(c[0]) + get_sz(c[1]);
+}
+void conn(node c, node p, int is_left) { 
+	if(c) c->p = p;
+	if(is_left>=0) p->c[!is_left] = c;
+}
+void rotate(node x){
+	node p = x->p, g = p->p;
+	bool gCh = p->is_root(), is_left = x==p->c[0];
+	conn(x->c[is_left],p,is_left); 
+	conn(p,x,!is_left);
+	conn(x,g,gCh?-1:(p==g->c[0])); 
+	p->upd();
+}
+void splay(node x){
+	while(!x->is_root()){
+		node p = x->p, g = p->p;
+		if(!p->is_root()) g->push();
+		p->push(); x->push();
+		if(!p->is_root()) rotate((x==p->c[0])==(p==g->c[0]) ? p : x);
+		rotate(x);
+	}
+	x->push(); x->upd();
+}
+
+// link cut tree
```

Comment:
Agreguemos una breve descripcion de que es y como esta compuesto esto internamente, cosa que sea mas facil de recordar en el futuro y entender en caso de que tengamos que tocar algo al usarlo. Algo asi como:

Keep information of a tree (or forest) and allow to make many types of operations (see them below) in an efficient way.
Internally, each node of the tree will have at most 1 "preferred" child, and as a consequence, the tree can be seen as a set of independent "preferred" paths. Each of this paths is basically a list, represented with a splay tree, where the "implicit key" (for the BST) of each element is the depth of the corresponding node in the original tree (or forest). Also, each of these preferred paths (except one of them), will know who its "father path" is, i.e. will know the preferred path of the father of the top-most node of this path.